### ISSUE TITLE
Add the ability to load overrides from multiple runtime config files and merge them

### ISSUE THREAD

**Issue by replay** [grafana/mimir#1798]
#### Is your feature request related to a problem? Please describe.

We are currently using scripts to merge overrides from multiple different sources and generate one single set of overrides, which then gets loaded by Mimir via the `-runtime-config.file` parameter. This requires external components to implement the merging logic, and to stay up2date if Mimir adds new properties to the overrides struct. Having this merge logic in external components introduces a potential source of errors which is outside the control of the Mimir project.

#### Describe the solution you'd like

It would be nice if Mimir could simply monitor multiple runtime config files and then merge them in the order in which they are specified, then we could remove the external potential source of errors (the merging logic).

#### Describe alternatives you've considered

Continue using external scripts to merge the overrides provided by different sources.

**Comment by pracucci** [comment]
> It would be nice if Mimir could simply monitor multiple runtime config files and then merge them in the order in which they are specified

👍  I think it makes sense.

**Comment by replay** [comment]
> What are the rules when two files specify the same limit at different values?

I would say "merge them in the order in which they are specified" and the last one wins, since each override overrides the previous value.

**Comment by replay** [comment]
We could make sure that the jsonnet-generated overrides are always the last ones to be merged, so they override all others in case of a conflict.

**Comment by stevesg** [comment]
Another option is to allow a particular value to only come from one of the overrides files. If a value is set by multiple files, then it is an error and we continue running with the current overrides.

This would mean the ordering of the files becomes irrelevant (less chance for mistakes), and one file could not accidentally change the value which it's not meant to.

Disadvantage being it limits the utility of the feature, but arguably makes it more robust.

**Comment by pracucci** [comment]
I like @stevesg idea, because:
1. It makes harder to shoot on your feet
2. It's better to be more strict at the beginning and eventually relax constraints in the future, than the opposite (the opposite would be a breaking change)

**Comment by pstibrany** [comment]
I think having single source for field can be tricky. In our case, we use a script to implement "free tier" overrides. (Ie. list of config files would be: 1) free tier script output, 2) custom overrides) But if such free-tier user asked for temporary increased limit, we would not be able to override "free tier" override.

**Comment by stevesg** [comment]
That particular script never produces overlapping overrides, it checks what is currently in the primary overrides and does not replace it.

**Comment by pstibrany** [comment]
> That particular script never produces overlapping overrides, it checks what is currently in the primary overrides and does not replace it.

Fair enough, although I would expect script could be simplified and drop the check, if it knew that limits can be overwritten later. I still think scenario I described isn't that unlikely to disallow it from the start.

**Comment by treid314** [comment]
In talking with Steve, I agree that we should reject fields with multiple overrides that would result in a merge and treat those cases like similar to a malformed config. I think merging configs adds more confusion for how these files merge and in what order the configs are merged. It's also a pretty rare case to have to merge fields for our operations today. Having the merge logic would be seldomly used and I feel add a lot of complexity to the task/product.

**Comment by pstibrany** [comment]
I think:

1) having the merge logic would be probably much easier to implement than rejecting overriden fields (just a gut-feeling – simply unmarshalling yamls in correct order into the same struct would do the trick)
2) we will need it sooner or later anyway, and I don't think complexity of "YAML files are ordered, and deserialized in this order" is very high.

However I'm fine starting with "reject" logic first, and only add "merging" when needed.
